core/markets/storage/network/libp2p_storage_market_network.cpp

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "markets/storage/network/libp2p_storage_market_network.hpp"
+#include "markets/storage/ask_protocol.hpp"
+#include "markets/storage/deal_protocol.hpp"
+#include "markets/storage/network/libp2p_ask_stream.hpp"
+#include "markets/storage/network/libp2p_deal_stream.hpp"
+
+namespace fc::markets::storage::network {
+
+  using libp2p::connection::Stream;
+  using libp2p::peer::PeerInfo;
+
+  outcome::result<std::shared_ptr<StorageAskStream>>
+  Libp2pStorageMarketNetwork::newAskStream(const PeerId &peer_id) {
+    PeerInfo peer_info = host_->getPeerRepository().getPeerInfo(peer_id);
+    std::shared_ptr<libp2p::connection::Stream> stream;
+    host_->newStream(
+        peer_info,
+        kAskProtocolId,
+        [&stream](outcome::result<std::shared_ptr<Stream>> new_stream) {
+          stream = std::move(new_stream.value());
+        });
+    return std::make_shared<Libp2pAskStream>(peer_id, stream);
+  }
+
+  outcome::result<std::shared_ptr<StorageDealStream>>
+  Libp2pStorageMarketNetwork::newDealStream(const PeerId &peer_id) {
```

Comment:
Connection creation is asynchronous and should use callback.
`outcome::result` can't return asynchronous errors and can't notify when connection is established.

Reply:
added callback handler